File: converted/utilities/string-similarity/elide-string-similarity.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-string-similarity

This demonstrates calling the TypeScript string similarity implementation from Python
using Elide's polyglot capabilities.

Benefits:
- One string matching implementation shared across TypeScript and Python
- Consistent fuzzy matching across services
- No Python similarity library needed
- Multiple algorithms in one package
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# similarity = require('./elide-string-similarity.ts')

print("=== Python Calling TypeScript String Similarity ===\n")

# Example 1: Basic similarity comparison
# similarity_score = similarity.compareTwoStrings("hello world", "hello word")
# print(f"Similarity: {similarity_score:.3f}")
# print()

# Example 2: Find best match for search
# search_term = "apple"
# options = ["apples", "banana", "app", "application", "pear"]
# result = similarity.findBestMatch(search_term, options)
# print(f"Search: '{search_term}'")
# print(f"Best match: '{result['bestMatch']['target']}' ({result['bestMatch']['rating']:.3f})")
# print()

# Example 3: Data deduplication in Python
# def find_duplicates(items, threshold=0.8):
#     duplicates = []
#     for i in range(len(items)):
#         for j in range(i + 1, len(items)):
#             score = similarity.compareTwoStrings(items[i], items[j])
#             if score >= threshold:
#                 duplicates.append((items[i], items[j], score))
#     return duplicates
#
# products = ["iPhone 13", "i-Phone 13", "Samsung Galaxy", "iPhone 13 Pro"]
# dupes = find_duplicates(products)
# print("Possible duplicates:")
# for item1, item2, score in dupes:
#     print(f"  '{item1}' ≈ '{item2}' ({score:.3f})")
# print()

# Example 4: Search autocomplete
# def autocomplete(query, database, limit=5):
#     results = similarity.findBestMatch(query, database)
#     matches = [r for r in results['ratings'] if r['rating'] > 0.3]
#     matches.sort(key=lambda x: x['rating'], reverse=True)
#     return matches[:limit]
#
# database = ["python", "javascript", "typescript", "java", "ruby", "rust", "go"]
# query = "javascrip"
# suggestions = autocomplete(query, database)
# print(f"Autocomplete for '{query}':")
# for match in suggestions:
#     print(f"  {match['target']}: {match['rating']:.2f}")
# print()

# Example 5: Name matching
# def match_customer(name, customer_database):
#     result = similarity.findBestMatch(name, customer_database)
#     best = result['bestMatch']
#     if best['rating'] > 0.7:
#         return best['target']
#     return None
#
# customers = ["John Smith", "Jane Doe", "Bob Johnson", "Alice Williams"]
# search = "Jon Smith"
# match = match_customer(search, customers)
# print(f"Looking for: '{search}'")
# print(f"Best match: '{match}'" if match else "No close match found")
# print()

print("Real-world use case:")
print("- Python API needs fuzzy search for product names")
print("- Uses same TypeScript implementation as Node.js frontend")
print("- Guarantees consistent matching logic across entire stack")
print("- No need to install Python similarity library")
print()

print("Example: E-commerce Search")
print("┌─────────────────────────────────────────┐")
print("│   Elide String Similarity (TypeScript) │")
print("│   elide-string-similarity.ts           │")
print("└─────────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │Frontend│          │  API   │")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same fuzzy matching everywhere!")
print()

print("Problem Solved:")
print("Before: Different similarity libs in TS and Python = inconsistent results")
print("After: One Elide implementation = identical matching across services")
print()

print("Performance Benefits:")
print("- Zero cold start overhead with Elide")
print("- Instant fuzzy matching")
print("- Shared runtime across languages")
print("- No Python library installation needed")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  similarity = require('./elide-string-similarity.ts')")
print("  score = similarity.compareTwoStrings('hello', 'hallo')")
print("  print(f'Similarity: {score}')  # That's it!")

File: conversions/string-similarity/elide-string-similarity.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-string-similarity
#
# This demonstrates calling the TypeScript string similarity implementation from Ruby
# using Elide's polyglot capabilities.
#
# Benefits:
# - One string matching implementation shared across TypeScript and Ruby
# - Consistent fuzzy matching across microservices
# - No Ruby similarity gem needed
# - Perfect for Rails, Sidekiq, background workers

puts "=== Ruby Calling TypeScript String Similarity ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# similarity = Elide.require('./elide-string-similarity.ts')

# Example 1: Basic similarity
# score = similarity.compareTwoStrings("hello world", "hello word")
# puts "Similarity: #{score.round(3)}"
# puts ""

# Example 2: Rails ActiveRecord duplicate detection
# class Product < ApplicationRecord
#   def self.find_duplicates(threshold = 0.8)
#     similarity = Elide.require('./elide-string-similarity.ts')
#     duplicates = []
#
#     all.each_with_index do |product1, i|
#       all[i+1..-1].each do |product2|
#         score = similarity.compareTwoStrings(product1.name, product2.name)
#         duplicates << [product1, product2, score] if score >= threshold
#       end
#     end
#
#     duplicates
#   end
# end

# Example 3: Sidekiq job for name matching
# class NameMatcherJob
#   include Sidekiq::Worker
#
#   def perform(search_name, candidate_names)
#     similarity = Elide.require('./elide-string-similarity.ts')
#     result = similarity.findBestMatch(search_name, candidate_names)
#
#     best_match = result['bestMatch']
#     puts "Searching for: #{search_name}"
#     puts "Best match: #{best_match['target']} (#{best_match['rating'].round(3)})"
#
#     # Process the match...
#   end
# end

# Example 4: Search suggestions
# def autocomplete_search(query, database, limit = 5)
#   similarity = Elide.require('./elide-string-similarity.ts')
#   result = similarity.findBestMatch(query, database)
#
#   result['ratings']
#     .select { |r| r['rating'] > 0.3 }
#     .sort_by { |r| -r['rating'] }
#     .take(limit)
# end
#
# database = ["ruby", "python", "javascript", "java", "rust", "go"]
# query = "javascrip"
# suggestions = autocomplete_search(query, database)
# puts "Autocomplete for '#{query}':"
# suggestions.each do |match|
#   puts "  #{match['target']}: #{match['rating'].round(2)}"
# end

# Example 5: Customer name normalization
# class Customer
#   def self.find_by_fuzzy_name(name, threshold = 0.7)
#     similarity = Elide.require('./elide-string-similarity.ts')
#     all_names = all.pluck(:name)
#     result = similarity.findBestMatch(name, all_names)
#
#     best = result['bestMatch']
#     return nil if best['rating'] < threshold
#
#     find_by(name: best['target'])
#   end
# end

puts "Real-world use case:"
puts "- Ruby background worker needs fuzzy matching for data cleanup"
puts "- Uses same TypeScript implementation as Node.js API"
puts "- Guarantees consistent matching logic across job queue"
puts "- No gem installation needed"
puts ""

puts "Example: Rails Application"
puts "┌─────────────────────────────────────────┐"
puts "│   Elide String Similarity (TypeScript) │"
puts "│   elide-string-similarity.ts           │"
puts "└─────────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Rails │  │Sidekiq │"
puts "    │  API   │  │  App   │  │ Worker │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same fuzzy matching everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Different gems in each service = inconsistent matching"
puts "After: One Elide implementation = 100% consistent results"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/similarity.rb"
puts "  SIMILARITY = Elide.require('./elide-string-similarity.ts')"
puts "  "
puts "  # app/models/product.rb"
puts "  class Product < ApplicationRecord"
puts "    def similar_products(threshold = 0.7)"
puts "      all_names = Product.pluck(:name)"
puts "      matches = SIMILARITY.findMatches(name, all_names, threshold)"
puts "      Product.where(name: matches.map { |m| m['target'] })"
puts "    end"
puts "  end"
puts ""

puts "Sidekiq Integration (when ready):"
puts "  class DuplicateDetectorJob"
puts "    include Sidekiq::Worker"
puts "    "
puts "    def perform"
puts "      similarity = Elide.require('./elide-string-similarity.ts')"
puts "      # Use for data deduplication"
puts "    end"
puts "  end"
